### PR TITLE
Rebase STI to support ONBUILD Docker images

### PR DESCRIPTION
**Pull request by mfojtik**

### CODE REVIEW

**Comment by mfojtik** [comment]
[test]

**Comment by soltysh** [comment]
LGTM

**Comment by mfojtik** [comment]
@bparees @smarterclayton I just tested this locally and the builder works. Here is my sample buildConfig:

https://gist.github.com/mfojtik/737a03990f0dcf102764

**Comment by mfojtik** [comment]
@bparees I think we should move out the `ContextDir` from the DockerBuilStrategy. I think the `ContextDir` is more an option for the `Source` than a strategy thing.

**Comment by soltysh** [comment]
@mfojtik now that we support `ContextDir` in all our strategies I'm :+1: 
I hope you've meant `BuildSource` struct :wink:

**Comment by bparees** [comment]
@mfojtik agree to making contextdir part of buildsource instead of strategy.

**Comment by mfojtik** [comment]
[test] again...

**Comment by mfojtik** [comment]
@bparees @smarterclayton PTAL

I added ContextDir into BuildSource, added conversion function and fixed tons of tests....

File: pkg/build/api/types.go [hunk 1]
```diff
@@ -83,6 +83,11 @@ const (
 type BuildSource struct {
 	Type BuildSourceType `json:"type,omitempty"`
 	Git  *GitBuildSource `json:"git,omitempty"`
+
+	// Specify the sub-directory where the source code for the application exists.
+	// This allows to have buildable sources in directory other than root of
+	// repository.
+	ContextDir string `json:"contextDir,omitempty"`
```
Comment (bparees) [inline pkg/build/api/types.go]: apparently we're not doing json tagging for the internal structs anymore.

Comment (mfojtik) [inline pkg/build/api/types.go]: @bparees in that case they should be all removed ;-) I will make the current build finish and remove the json tagging.

**Comment by bparees** [comment]
i'm not seeing anything that actually tests this behavior, am i missing it?

**Comment by smarterclayton** [comment]
+1

----- Original Message -----

**Comment by bparees** [comment]
by "this behavior" i mean the contextdir part, i know you cover onbuild tests in STI.

**Comment by mfojtik** [comment]
@bparees the contextDir testing really belongs to STI as the Origin is only consuming it. There are plenty of tests in Origin where we do exercise contextDir option in BuildSource.

Also I just noted that there are no unit tests for onbuild. I going to add those into STI as a follow up.

**Comment by bparees** [comment]
@mfojtik something needs to test the Source.contextDir field behavior.

**Comment by bparees** [comment]
As discussed on IRC, cover contextDir behavior testing via e2e.

**Comment by mfojtik** [comment]
@bparees @smarterclayton all comments addressed, I think this is OK for merge. 

I'm still thinking about how to exercise this in e2e. I don't want to add example ruby app into sub-folder just to trigger ContextDir... Anyway, I think we can do this as a follow up for this card.

**Comment by bparees** [comment]
lgtm.

**Comment by mfojtik** [comment]
[merge]

**Pull request merged**